compiler/rustc_borrowck/src/type_check/mod.rs

```diff
@@ -2120,8 +2119,9 @@ impl<'a, 'tcx> TypeChecker<'a, 'tcx> {
                                 //
                                 // Note that other checks (such as denying `dyn Send` -> `dyn
                                 // Debug`) are in `rustc_hir_typeck`.
-                                if let ty::Dynamic(src_tty, ..) = src_tail.kind()
-                                    && let ty::Dynamic(dst_tty, ..) = dst_tail.kind()
+                                if let ty::Dynamic(src_tty, src_region_bound, ..) = src_tail.kind()
+                                    && let ty::Dynamic(dst_tty, dst_region_bound, ..) =
+                                        dst_tail.kind()
                                     && src_tty.principal().is_some()
                                     && dst_tty.principal().is_some()
```

Comment:
Is the "when there aren't principal traits" still true? It seems like I could get into trouble with a `fn (self: dyn Send + 'static)`, right?

Reply:
We forbid casting `dyn Send` to `dyn Trait` in general as we would have no way to construct a vtable. The opposite cast is also fine as the vtable for `dyn Send` (or other autotraits) necessarily doesn't contain potentially user-written functions with where clauses that could rely on the dyn type's lifetime bound. I should write this rationale down somewhere